File: tests/test_auth.py
from fastapi.testclient import TestClient
from src.core.config import settings

def test_register_user(client: TestClient):
    response = client.post(
        f"{settings.API_V1_STR}/auth/register",
        json={
            "email": "test@example.com",
            "password": "Password123!",
            "full_name": "Test User",
            "tenant_name": "Test Tenant"
        },
    )
    assert response.status_code == 200
    assert response.json()["email"] == "test@example.com"
    assert "tenant_id" in response.json()

def test_login_user(client: TestClient):
    # Primeiro registra
    client.post(
        f"{settings.API_V1_STR}/auth/register",
        json={
            "email": "login@example.com",
            "password": "Password123!",
            "full_name": "Login User",
            "tenant_name": "Login Tenant"
        },
    )
    
    # Tenta Login
    response = client.post(
        f"{settings.API_V1_STR}/auth/login",
        data={"username": "login@example.com", "password": "Password123!"},
    )
    assert response.status_code == 200
    assert "access_token" in response.json()
    assert response.json()["token_type"] == "bearer"

def test_get_me(client: TestClient):
    # Registro e Login
    email = "me@example.com"
    client.post(
        f"{settings.API_V1_STR}/auth/register",
        json={
            "email": email,
            "password": "Password123!",
            "full_name": "Me User",
            "tenant_name": "Me Tenant"
        },
    )
    login_res = client.post(
        f"{settings.API_V1_STR}/auth/login",
        data={"username": email, "password": "Password123!"},
    )
    token = login_res.json()["access_token"]
    
    # Acesso rota protegida
    response = client.get(
        f"{settings.API_V1_STR}/auth/me",
        headers={"Authorization": f"Bearer {token}"}
    )
    assert response.status_code == 200
    assert response.json()["email"] == email

def test_login_invalid_password(client: TestClient):
    email = "wrong@example.com"
    client.post(
        f"{settings.API_V1_STR}/auth/register",
        json={
            "email": email,
            "password": "Password123!",
            "full_name": "Wrong User",
            "tenant_name": "Wrong Tenant"
        },
    )
    response = client.post(
        f"{settings.API_V1_STR}/auth/login",
        data={"username": email, "password": "WrongPassword"},
    )
    assert response.status_code == 400
    assert response.json()["detail"] == "Incorrect email or password"

File: node-version/test_auth.js
const axios = require('axios');

const BASE_URL = 'http://localhost:8001/api/v1';

// Dados de teste para o login
const testUser = {
  name: "Auth Test User",
  email: `auth_test_${Date.now()}@example.com`,
  password: "Password123!",
  tenant_name: `Tenant Auth ${Date.now()}`
};

async function runAuthTests() {
  console.log('🔐 Iniciando bateria de testes de Autenticação...\n');

  try {
    // 1. Cadastrar Usuário
    console.log(`[1] Registrando usuário: ${testUser.email}...`);
    const registerRes = await axios.post(`${BASE_URL}/auth/register`, testUser);
    console.log('✅ Cadastro realizado com sucesso!');
    console.log('   Dados do retorno:', registerRes.data);

    // 2. Fazer Login
    console.log(`\n[2] Realizando Login com as credenciais criadas...`);
    const loginRes = await axios.post(`${BASE_URL}/auth/login`, {
      email: testUser.email,
      password: testUser.password
    });
    
    // Agora o login devolve só access_token e refresh_token (seguro)
    const { access_token, refresh_token } = loginRes.data;
    console.log('✅ Login realizado com sucesso! (Sem vazamento de tenant/user ID)');
    console.log(`   Access Token JWT : ${access_token.substring(0, 30)}...`);
    console.log(`   Refresh Token JWT: ${refresh_token.substring(0, 30)}...`);

    // 3. Validação do Token (Endpoint Me)
    console.log(`\n[3] Validando o access_token acessando uma rota protegida (/auth/me)...`);
    const meRes = await axios.get(`${BASE_URL}/auth/me`, {
      headers: { Authorization: `Bearer ${access_token}` }
    });
    console.log('✅ Acesso autorizado e validado!');
    console.log(`   Perfil do usuário autenticado retornado: Tenant=${meRes.data.tenant_id}, ID=${meRes.data.id}`);

    // 4. Rotacionamento de Token (Refresh)
    console.log(`\n[4] Solicitando Rotacionamento de Token (/auth/refresh)...`);
    const refreshRes = await axios.post(`${BASE_URL}/auth/refresh`, {
      refresh_token: refresh_token
    });
    console.log('✅ Novo par de tokens gerado com sucesso!');
    console.log(`   NOVO Access Token : ${refreshRes.data.access_token.substring(0, 30)}...`);
    console.log(`   NOVO Refresh Token: ${refreshRes.data.refresh_token.substring(0, 30)}...`);

    console.log('\n🎉 Todos os testes de Autenticação (Registro -> Login -> Validação JWT) passaram!');
  } catch (error) {
    console.error('\n❌ ERRO FATAL no teste de Autenticação:');
    if (error.response) {
      console.error(`Status: ${error.response.status}`);
      console.error(error.response.data);
    } else {
      console.error(error.message);
    }
  }
}

runAuthTests();
